CMakeLists.txt

```diff
@@ -26,82 +24,162 @@ if(HIOP_USE_MPI)
     find_package(MPI REQUIRED)
     set(CMAKE_CXX_COMPILER ${MPI_CXX_COMPILER})
   endif(NOT DEFINED MPI_CXX_COMPILER)
-  file(APPEND "src/Interface/hiop_defs.hpp" "#define HIOP_USE_MPI\n")
-else()
-  file(APPEND "src/Interface/hiop_defs.hpp" "#ifdef HIOP_USE_MPI\n#undef HIOP_USE_MPI\n#endif\n")
 endif(HIOP_USE_MPI)
 
+# The binary dir is already a global include directory
+configure_file(
+  "${CMAKE_SOURCE_DIR}/src/Interface/hiop_defs.hpp.in"
+  "${CMAKE_BINARY_DIR}/hiop_defs.hpp")
+
+add_library(hiop_magma INTERFACE)
+add_library(hiop_math INTERFACE)
+add_library(hiop_cuda INTERFACE)
+add_library(hiop_options INTERFACE)
+add_library(hiop_warnings INTERFACE)
+
+target_compile_features(hiop_options
+    INTERFACE
+        cxx_std_11
+        cxx_alignas
+        cxx_alignof
+        cxx_attributes
+        cxx_auto_type
+        cxx_constexpr
+        cxx_defaulted_functions
+        cxx_deleted_functions
+        cxx_final
+        cxx_lambdas
+        cxx_noexcept
+        cxx_override
+        cxx_range_for
+        cxx_rvalue_references
+        cxx_static_assert
+        cxx_strong_enums
+        cxx_trailing_return_types
+        cxx_unicode_literals
+        cxx_user_literals
+        cxx_variadic_macros
+)
+
+target_compile_options(hiop_warnings
```

Comment:
If you really like enforcing these compile options (I also kinda like them :)), can I suggest to have them enabled only under a "developer mode"?

cmake -DHIOP_DEVELOPER_MODE=ON ..

I guess HIOP_DEVELOPER_MODE should be off by default, to not disturb the users

Reply:
I think this is a great idea!